File: Java Project/Iteration_2/src/CourseRegistrationSimulation.java
package Iteration_2.src;


import org.apache.log4j.Logger;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class CourseRegistrationSimulation {
    private ArrayList<Student> students;
    private ArrayList<CompulsoryCourse> courses;
    private ArrayList<Lecturer> lecturers;
    private ArrayList<Advisor> advisors;
    private int creditLimit;
    private ArrayList<EnrollmentRequest> enrollmentRequests;
    private ArrayList<String> errors;
    private double probToPassClass;
    private double probTo;
    private String[] names = {"Ahmet", "Ali", "Ayşe", "Fatma", "Kemal"};
    private String[] surnames = {"Kebapçı", "Çevik", "Öztürk", "Vural", "Ertekin"};
    private String[] letterGrades = {"AA", "BA", "BB", "CB", "CC", "DC"};
    private String[] year = {"2019", "2020", "2021"};
    private String term;
    private int studentSize;
    private Logger logger = Logger.getLogger(CourseRegistrationSimulation.class.getName());


    public CourseRegistrationSimulation(ArrayList<Student> students, ArrayList<CompulsoryCourse> courses, ArrayList<Lecturer> lecturers, ArrayList<Advisor> advisors, int creditLimit) {
        this.students = students;
        this.courses = courses;
        this.lecturers = lecturers;
        this.advisors = advisors;
        this.creditLimit = creditLimit;
    }

    public CourseRegistrationSimulation() {
    }

    public void starSimulation() throws IOException {

        logger.info("Simulation Started. ");
        FileManager1 fileManager1 = new FileManager1();
        createParams();

        createLecturer();

        createAdvisors();
        createCourses();

        students = createRandomStudent(studentSize);
        for (Student value : students)
            fileManager1.writeToFile(value, "Java Project/Iteration_2/src/Jsons/Students/" + value.getStudentID() + ".json");

        createStudents();
        matchStudentAdvisor();
        for (Student student : this.students) {
            ArrayList<CompulsoryCourse> appliedCourses = applyCourse(student);
            EnrollmentRequest enrollmentRequest = new EnrollmentRequest(appliedCourses, student);
            logger.info("Student "+ student.getStudentID()+" created new Enrollement Requests with courses "+appliedCourses.stream().map(CompulsoryCourse::getCourseCode).toList());
            Advisor advisor = student.getAdvisor();
            checkSystemRequirements(enrollmentRequest);
            advisor.checkScheduleCollision(enrollmentRequest);
            logger.info("Added Courses to Student "+student.getStudentID()+" : "+ enrollmentRequest.getCourses().stream().map(CompulsoryCourse::getCourseCode).toList());
            student.calculateTranscriptAfter(enrollmentRequest,this.probToPassClass);
            fileManager1.writeToFile(student,("Java Project/Iteration_2/src/Outputs/"+ student.getStudentID()+".json"));
            try {
                TimeUnit.MILLISECONDS.sleep(300);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

        }
        logger.info("Simulation is over.");


    }


    public void createCourses() {
        FileManager1 fileManager1 = new FileManager1();
        this.courses = fileManager1.readCourse("Java Project/Iteration_2/src/Jsons/compulsoryCourses.json");
        logger.info(this.courses.size()+" Course Created.");

    }

    public void createStudents() throws IOException {//done
        FileManager1 fileManager1 = new FileManager1();
        ArrayList<Student> students = fileManager1.readStudent("Java Project/Iteration_2/src/Jsons/Students");

        this.students = students;

    }
    public void createParams(){
        FileManager1 fileManager1 = new FileManager1();
        HashMap<String, String> params = fileManager1.readParams("Java Project/Iteration_2/src/input.json") ;
        this.probToPassClass =  Double.parseDouble(params.get("prob_to_fail_class"));
        this.term =  params.get("semester");
        double number = Double.parseDouble(params.get("student_count"));
        this.studentSize = (int)(number);


    }

    public void createLecturer() {
        FileManager1 fileManager1 = new FileManager1();
        this.lecturers = fileManager1.readLecturers("Java Project/Iteration_2/src/Jsons/lecturer.json");

    }

    public void createAdvisors() {
        Lecturer lecturer = new Lecturer();
        this.advisors = lecturer.lecturerToAdvisor(lecturers);
        logger.info("Advisors created from Lecturers.");
    }

    /*public void createAdvisor() {  // no need for iteration 1
        FileManager fileManager = new FileManager();
        this.advisors = fileManager.readAdvisor("/Users/omerfarukbulut/Downloads/student1.json");
    }*/

    public void matchStudentAdvisor() {
        Random random = new Random();
        int size = advisors.size();
        for (Student student : this.students) {
            Advisor advisor = this.advisors.get(random.nextInt(size));
            student.setAdvisor(advisor);
            advisor.appendStudent(student.getStudentID());
            logger.info("Student "+student.getStudentID()+" match with "+advisor.getLecturerName()+" "+advisor.getLecturerSurname());

        }
    }

    public ArrayList<CompulsoryCourse> applyCourse(Student student) {
        ArrayList<CompulsoryCourse> courses = this.courses;
        Random random = new Random();
        int appliedCourseNumber = random.nextInt(1, 10);
        ArrayList<CompulsoryCourse> appliedCourses = new ArrayList<CompulsoryCourse>();
        for (int i = 0; i < appliedCourseNumber; i++) {
            CompulsoryCourse randomCourse = courses.get(random.nextInt(courses.size()));
            appliedCourses.add(randomCourse);
            System.out.println(student.getStudentID() + " applied for " + randomCourse.getCourseCode());
            logger.info(student.getStudentID() + " applied for " + randomCourse.getCourseCode());
            // Log


        }

        return appliedCourses;


    }

    public void checkSystemRequirements(EnrollmentRequest enrollmentRequest) {
        logger.info("System Checking "+enrollmentRequest.getStudent().getStudentID()+"'s enrollment requests");
        RegistrationSystem registrationSystem = new RegistrationSystem();
        registrationSystem.checkCourseIsTakenBefore(enrollmentRequest);
        registrationSystem.checkPrerequisites(enrollmentRequest);
        registrationSystem.isOpenedThisTerm(enrollmentRequest, this.term);

    }


    public ArrayList<CompulsoryCourse> getRandomCourse(ArrayList<CompulsoryCourse> list) {
        Random random = new Random();
        ArrayList<CompulsoryCourse> newList = new ArrayList<CompulsoryCourse>();

        for (int i = 0; i < random.nextInt(list.size()); i++) {
            int randomIndex = random.nextInt(list.size());
            newList.add(list.get(randomIndex));

        }
        return newList;
    }

    public boolean passOrFail() {
        Random random = new Random();
        return random.nextDouble() < this.probToPassClass;
    }


    public ArrayList<Student> createRandomStudent(int size) {
        ArrayList<Student> students1 = new ArrayList<Student>();
        for (int i = 0; i < size; i++) {
            students1.add(new Student(names[new Random().nextInt(names.length)], surnames[new Random().nextInt(surnames.length)], new Transcript(), new Transcript(), new Schedule(), new Advisor(), year[new Random().nextInt(year.length)]));
            students1.get(i).createStudentID();

            for(int k = 0; k < students1.size()-1; k++) {
                if(students1.get(i).getStudentID().equals(students1.get(k).getStudentID())) {
                    students1.get(i).createStudentID();
                    k = 0;
                }
            }

            Transcript transcriptBefore = students1.get(i).getTranscriptBefore();
            transcriptBefore.setTakenCourses(getRandomCourse(courses));

            HashMap<String, String> courseGrades = new HashMap<String, String>();
            int takenCredit = 0;
            for (int j = 0; j < transcriptBefore.getTakenCourses().size(); j++) {
                takenCredit += transcriptBefore.getTakenCourses().get(j).getCredit();
                transcriptBefore.setTakenCredit(takenCredit);

                courseGrades.put(transcriptBefore.getTakenCourses().get(j).getCourseCode(), letterGrades[new Random().nextInt(letterGrades.length)]);
            }
            transcriptBefore.setCourseGrades(courseGrades);

            transcriptBefore.setGpa(transcriptBefore.calculateGpa());

            students1.get(i).setAdvisor(advisors.get(new Random().nextInt(advisors.size())));


        }
        return students1;
    }

//    public void writeToJson() {
//    	JsonArrayBuilder builder = Json.createArrayBuilder();
//
//    	for(int i = 0; i < students.size(); i++) {
//        		     builder.add(Json.createObjectBuilder()
//        		         .add("studentID", students.get(i).getStudentID())
//        		         .add("studentName", students.get(i).getStudentName())
//        		         .add("studentSurname", students.get(i).getStudentSurname())
//        		         .add("transcriptBefore", students.get(i).getTranscriptBefore().toString())
//        		         .add("transcriptAfter", students.get(i).getTranscriptAfter().toString())
//        		         .add("schedule", students.get(i).getSchedule().toString())
//        		         .add("advisor", students.get(i).getAdvisor().toString())
//        		         .add("enrolledYear", students.get(i).getEnrolledYear()))
//        		     .build();
//    	}
//    	//JsonArray array = builder.build();
//
//    	try {
//    		JsonWriter writer = (JsonWriter) Json.createWriter(new FileOutputStream("deneme.json"));
//    		//writer.writeArray(array);
//    		writer.close();
//
//    	} catch (IOException e) {
//    		e.printStackTrace();
//    	}
//    }

/*
    public ArrayList<ElectiveCourse> getRandomElectiveCourses(ArrayList<ElectiveCourse> list) {
        Random rand = new Random(); // object of Random class.

        //temprory list to hold selected items.
        ArrayList<ElectiveCourse> tempList = new ArrayList<>();
        for (int i = 0; i < rand.nextInt(list.size()); i++) {
            int randomIndex = rand.nextInt(list.size());
            tempList.add(list.get(randomIndex));
        }
        return tempList;
    }*/

    public void printArraylist(ArrayList<Student> arr) {
        for (int i = 0; i < arr.size(); i++) {
            // Assuming arr is an ArrayList object
            Object current = arr.get(i).toString();
            System.out.println(current);
            // Do something with the current element here
        }

    }

    public ArrayList<String> getErrors() {
        return errors;
    }

    public void setErrors(ArrayList<String> errors) {
        this.errors = errors;
    }

    /*public void checkCourseIsOpened() {
        Semester semester = this.semester;
        for (EnrollmentRequest enrollmentRequest : this.enrollmentRequests) {
            for (CompulsoryCourse course : enrollmentRequest.getCourses()) {
                if (!semester.getOpenedCourse().contains(course)) {

                    enrollmentRequest.setError(enrollmentRequest.getStudent().getStudentName() + course.getCourseCode() + " not opened");
                    ArrayList<String> errorsBefore = new ArrayList<String>();
                    errorsBefore.add(enrollmentRequest.getStudent().getStudentName() + course.getCourseCode() + " not opened");
                    this.errors = errorsBefore;
                }

            }
        }

    }*/
}

File: Python/src/CourseRegistrationSimulation.py
import logging
import time
import random
from typing import List

import Student
import CompulsoryCourse
import Lecturer
import Advisor
import EnrollmentRequest
import FileManager1
import RegistrationSystem

logger = logging.getLogger(__name__)

class CourseRegistrationSimulation:

    __students: List[Student]
    __courses: List[CompulsoryCourse]
    __lecturers: List[Lecturer]
    __advisors: List[Advisor]
    __credit_limit: int
    __enrollment_requests: List[EnrollmentRequest]
    __errors: List[str]
    __prob_to_pass_class: float
    __prob_to: float
    __names = ["Ahmet", "Ali", "Ayşe", "Fatma", "Kemal"]
    __surnames = ["Kebapçı", "Çevik", "Öztürk", "Vural", "Ertekin"]
    __letter_grades = ["AA", "BA", "BB", "CB", "CC", "DC"]
    __year = ["2019", "2020", "2021"]
    __term: str
    __student_size: int
    
    def __init__(self, students: List[Student], courses: List[CompulsoryCourse], lecturers: List[Lecturer], advisors: List[Advisor], credit_limit: int):
        self.__students = students
        self.__courses = courses
        self.__lecturers = lecturers
        self.__advisors = advisors
        self.__credit_limit = credit_limit   

    def star_simulation(self):
        #self.logger.info("Simulation Started.")
        file_manager = FileManager1.FileManager1()
        self.create_params()
        self.create_lecturers()
        self.create_advisors()
        self.create_courses()
        self.__students = self.create_random_students(self.student_size)
        for student in self.__students:
            file_manager.write_to_file(student, f"Jsons/Students/{student.student_id}.json")
        self.create_students()
        self.match_student_advisor()
        for student in self.__students:
            applied_courses = self.apply_course(student)
            enrollment_request = EnrollmentRequest(applied_courses, student)
            self.logger.info(f"Student {student.student_id} created new Enrollment Requests with courses {[c.course_code for c in applied_courses]}")
            advisor = student.advisor
            self.check_system_requirements(enrollment_request)
            advisor.check_schedule_collision(enrollment_request)
            self.logger.info(f"Added courses to Student {student.student_id}: {[c.course_code for c in enrollment_request.courses]}")
            student.calculate_transcript_after(enrollment_request, self.prob_to_pass_class)
            file_manager.write_to_file(student, f"Outputs/{student.student_id}.json")
            time.sleep(0.3)
        self.logger.info("Simulation is over.")
    
    def create_courses(self):
        file_manager = FileManager1.FileManager1()
        self.__courses = file_manager.read_course("Jsons/compulsoryCourses.json")
        self.logger.info(len(self.__courses)+" Course Created.")

    def create_students(self):
        file_manager = FileManager1.FileManager1()
        students = file_manager.read_student("Jsons/Students")
        self.__students = students

    def create_params(self):
        try:
            file_manager = FileManager1.FileManager1()
            params = file_manager.read_params("input.json")
            self.__prob_to_pass_class = float(params["prob_to_fail_class"])
            self.__term = params["semester"]
            number = float(params["student_count"])
            self.__student_size = int(number)
        except Exception as e:
            print(e)
        

    def create_lecturers(self):
        file_manager = FileManager1.FileManager1()
        self.__lecturers = file_manager.read_lecturers("Jsons/lecturer.json")

    def create_advisors(self):
        print(self.__lecturers)
        advisor = Advisor.Advisor(None, None, None, None)
        self.__advisors = advisor.lecturer_to_advisor(self.__lecturers)
        logger.info("Advisors created from Lecturers.")

    def match_student_advisor(self):
        size = len(self.__advisors)
        for student in self.__students:
            advisor = self.__advisors[random.randint(0, size-1)]
            student.set_advisor(advisor)
            advisor.append_student(student.get_student_id())
            logger.info(f"Student {student.get_student_id()} match with {advisor.get_lecturer_name()} {advisor.get_lecturer_surname()}")

    def applyCourse(self, student):
        appliedCourseNumber = random.randint(1, 10)
        appliedCourses = []
        for i in range(appliedCourseNumber):
            randomCourse = self.__courses[random.randint(len(self.courses))]
            appliedCourses.append(randomCourse)
            logger.info(f"{student.getStudentID()} applied for {randomCourse.getCourseCode()}")
        return appliedCourses

    def check_system_requirements(self, enrollment_request):
        logger.info("System Checking {}'s enrollment requests".format(enrollment_request.get_student().get_student_id()))
        registration_system = RegistrationSystem()
        registration_system.check_course_is_taken_before(enrollment_request)
        registration_system.check_prerequisites(enrollment_request)
        registration_system.is_opened_this_term(enrollment_request, self.__term)

    def get_random_course(list):
        new_list = []

        for i in range(random.randint(0, len(list))):
            random_index = random.randint(0, len(list))
            new_list.append(list[random_index])

        return new_list

    def pass_or_fail(self):
        return random.random() < self.__prob_to_pass_class

    def create_random_student(self, size):
        students = []
        for i in range(size):
            students.append(Student(self.__names[random.randint(0, len(self.__names))], self.__surnames[random.randint(0, len(self.__surnames))], Transcript(), Transcript(), Schedule(), Advisor(), self.__year[random.randint(0, len(self.__year))]))
            students[i].create_student_id()

            for k in range(len(students) - 1):
                if students[i].get_student_id() == students[k].get_student_id():
                    students[i].create_student_id()
                    k = 0

            transcript_before = students[i].get_transcript_before()
            transcript_before.set_taken_courses(self.get_random_course(self.__courses))

            course_grades = {}
            taken_credit = 0
            for j in range(len(transcript_before.get_taken_courses())):
                taken_credit += transcript_before.get_taken_courses()[j].get_credit()
                transcript_before.set_taken_credit(taken_credit)

                course_grades[transcript_before.get_taken_courses()[j].get_course_code()] = self.__letter_grades[random.randint(0, len(self.__letter_grades))]
            transcript_before.set_course_grades(course_grades)

            transcript_before.set_gpa(transcript_before.calculate_gpa())

            students[i].set_advisor(self.__advisors[random.randint(0, len(self.__advisors))])

        return students

    def print_arraylist(arr):
        for i in range(len(arr)):
            # Assuming arr is a list object
            current = arr[i].to_string()
            print(current)
            # Do something with the current element here

    def get_errors(self):
        return self.__errors

    def set_errors(self, errors):
        self.__errors = errors

